Написал API тесты для учебного сервиса Яндекс Самокат

CRASH3000 opened the pull request:
1) Созданы тесты для проверки API:
  - Создание курьера
  - Логин курьера
  - Создание заказа
  - Получение списка заказов
  - Удаление курьера (дополнительное задание)

2) Настроен проект с использованием pytest и Allure для генерации отчетов. 
3) Добавлены вспомогательные модули и данные для тестов. 
4) Создан README с описанием проекта и инструкциями по запуску. 
5) Добавлен файл requirements.txt с зависимостями проекта.

✅ Все автотесты проверены и работают. 
⚠ Кроме того, тесты обнаруживают ошибки, а именно несоответствие ожидаемых ответов на запросы указанные в  ТЗ (Swagger) с фактическим результатом. Подробнее можно увидеть в Allure отчете.

![Снимок экрана 2024-09-01 в 21 17 14](https://github.com/user-attachments/assets/3617f986-2da8-47ae-bd1b-cccdecd8e493)
![Снимок экрана 2024-09-01 в 21 17 49](https://github.com/user-attachments/assets/ad209b29-d5e2-4893-aae4-e120d097af47)

CRASH3000 committed:
Написал API тесты для учебного сервиса Яндекс Самокат

1) Созданы тесты для проверки API:
  - Создание курьера
  - Логин курьера
  - Создание заказа
  - Получение списка заказов
  - Удаление курьера (дополнительное задание)

2) Настроен проект с использованием pytest и Allure для генерации отчетов.
3) Добавлены вспомогательные модули и данные для тестов.
4) Создан README с описанием проекта и инструкциями по запуску.
5) Добавлен файл requirements.txt с зависимостями проекта.

CRASH3000 committed:
Небольшой фикс:
- Удалил из .gitignore игнорирование allure-results/
- Добавил папку allure-results в репозиторий c актуальным прогоном тестов.

AlexKitIt commented on API-tests/test_scooter_api.py:
```diff
@@ -0,0 +1,386 @@
+        """Метод, который выполняется один раз перед запуском всех тестов в классе"""
+        cls.courier_data = generate_random_courier_data()
+
+    @allure.feature("Создание курьера")
+    @allure.story("1. Создание курьера")
```
Нужно исправить: здесь и далее: эти декораторы ставятся над тестовым классом, над тестами ставится title

CRASH3000 replied:
✅ Поправил. Перенес @allure.feature и @allure.story над тестовые классы и в сами тесты добавил @allure.title

AlexKitIt commented on API-tests/test_scooter_api.py:
```diff
@@ -0,0 +1,386 @@
+        assert response.json() == {
+            "ok": True
+        }, "Первичный запрос на создание курьера вернул некорректный ответ"
+        print(f"Курьер создан: {self.courier_data}")
```
Нужно исправить: принтов быть не должно

AlexKitIt commented on API-tests/test_scooter_api.py:
```diff
@@ -0,0 +1,386 @@
+    )
+    def test_create_courier(self):
+        response = requests.post(f"{Url.BASE_URL}/courier", json=self.courier_data)
+        print(f"Статус после создания курьера в setup_class: {response.status_code}")
```
Нужно исправить: здесь и далее: принтов быть не должно

CRASH3000 replied:
✅ Поправил. Принты очень помогали понять, что приходит ответом как в постмане. Принты удалил из тестов.

AlexKitIt replied:
При дебаге вопросов нет, но в прод они попадать не должны

AlexKitIt commented on API-tests/test_scooter_api.py:
```diff
@@ -0,0 +1,386 @@
+
+        assert response.status_code == 409, "Код ответа не 409 для повторного логина"
+        assert (
+            response.json().get("message") == "Этот логин уже используется"
```
Можно лучше: тексты ответов стоит хранить снаружи в data модулях для поддержания и актуализации

CRASH3000 replied:
✅ Поправил. В папке data, создал отдельный файл для хранения ответов expected_responses.py. 
Импортировал содержимое файла в тесты.

AlexKitIt commented on API-tests/test_scooter_api.py:
```diff
@@ -0,0 +1,386 @@
+        ), "Сообщение об ошибке не совпадает"
+
+
+class TestCourierLoginAPI:
```
Нужно исправить: здесь и далее: разные классы стоит разделить по разным модулям

CRASH3000 replied:
✅ Поправил. Теперь каждый класс с набором тестов хранится в отдельном модуле.

AlexKitIt commented on API-tests/test_scooter_api.py:
```diff
@@ -0,0 +1,386 @@
+    def setup_class(cls):
+        """Используем данные из TestCourierAPI"""
+        # Убедимся, что курьер был создан
+        TestCourierAPI.setup_class()  # Явно вызываем setup_class из TestCourierAPI
```
Нужно исправить: здесь и далее: методы тестовых классов не вызываются декларативно. Лучше для всего этого использовать фистуры

CRASH3000 replied:
✅ Самая сложная задачка, но вроде разобрался. В папке с тестами создал новый файл conftest.py и перенес логику  @classmethod в фистуры.

CRASH3000 committed:
API тесты для учебного сервиса Яндекс Самокат v2.0

Отработал все замечания ревьюера по API тестам.

Что нового?
- Создал отдельные модули для каждого класса с набором тестов;
- В папке с тестами создал новый файл conftest.py и перенес логику @classmethod в фистуры;
- В папке data, создал отдельный файл для хранения ответов expected_responses.py и
импортировал содержимое файла в тесты.

Внес правки в код.
- Перенес @allure.feature и @allure.story над тестовые классы и в сами тесты добавил @allure.title.
- Заменил @allure.feature на @allure.suite;
- Сделал код проекта и Allure отчет более чистым;
- Удалил все принты из тестов.

CRASH3000 commented:
Повторно проверил тесты API, после правок кода.
✅  Все автотесты работают.
![Снимок экрана 2024-09-04 в 02 50 50](https://github.com/user-attachments/assets/b24a9e08-3a72-4e2a-a732-38032fc25a58)
![Снимок экрана 2024-09-04 в 02 50 38](https://github.com/user-attachments/assets/e59feb94-614d-4626-90b6-da82a454b105)

AlexKitIt commented on API-tests/conftest.py:
```diff
@@ -0,0 +1,23 @@
+from data.courier_generator_data import generate_random_courier_data
+from data.url_data import Url
+
+@pytest.fixture(scope="class")
```
Можно лучше: курьера лучше создавать каждый тест нового

CRASH3000 replied:
✅  Переписал логику, теперь при запуске каждого теста, создается новый курьер.

AlexKitIt commented on API-tests/conftest.py:
```diff
@@ -0,0 +1,23 @@
+def create_courier():
+    courier_data = generate_random_courier_data()
+    response = requests.post(f"{Url.BASE_URL}/courier", json=courier_data)
+    assert response.status_code == 201, "Не удалось создать курьера в фикстуре"
```
Нужно исправить: здесь и далее: фикстуры не тесты, в них assert не используется

CRASH3000 replied:
✅ Удалил assert во всех фикстурах.

AlexKitIt commented on API-tests/conftest.py:
```diff
@@ -0,0 +1,23 @@
+    courier_data = generate_random_courier_data()
+    response = requests.post(f"{Url.BASE_URL}/courier", json=courier_data)
+    assert response.status_code == 201, "Не удалось создать курьера в фикстуре"
+    return courier_data, response.json()
```
Можно лучше: сюда лучше прикрутить удаление, чтобы в БД мусора не было

CRASH3000 replied:
✅  Добавил логику в фикстурах, где после завершения теста, созданный тестовый профиль курьера удаляется.

AlexKitIt commented on API-tests/test_01_create_courier_api.py:
```diff
@@ -0,0 +1,135 @@
+
+@allure.suite("1. Создание курьера")
+@allure.story("Тестирование функционала создания курьера")
+@pytest.mark.usefixtures("create_courier")
```
Нужно исправить: здесь и далее: фикстура передается на вход тестового метода, тут эта конструкция тогда не нужна

CRASH3000 replied:
✅ Да, не увидел этот дубль. Везде убрал.

AlexKitIt commented on API-tests/test_01_create_courier_api.py:
```diff
@@ -0,0 +1,135 @@
+        "Отправляем POST-запрос на создание курьера с сгенерированными данными в Json файле"
+    )
+    def test_create_courier(self, create_courier):
+        courier_data, response_data = create_courier
```
Не очень хорошо что весь тест выполнился в предусловии. В тестах регистрации стоит описывать создание как есть

CRASH3000 replied:
✅ Убрал.

AlexKitIt commented on API-tests/test_01_create_courier_api.py:
```diff
@@ -0,0 +1,135 @@
+    @allure.description(
+        "Проверка, что курьера можно создать и ответом приходит код 201 и 'ok': True"
+    )
+    @allure.step(
```
Нужно исправить: здесь и далее: step ставится над логическими методами используемыми в тестах, над тестовыми методами только title

CRASH3000 replied:
✅  Убрал, но требуется дополнительное пояснение, по чему так.
Я добавил шаги, чтобы отчет в Аллюре был максимально заполнен и читаем.

AlexKitIt replied:
Когда ты описываешь методы - шаги (в твоем случае в модуле api это именно шаги и их несколько модет быть в тесте, потому step, а title  - это название теста)

AlexKitIt commented on API-tests/test_01_create_courier_api.py:
```diff
@@ -0,0 +1,135 @@
+    @allure.step(
+        "Отправляем POST-запрос на создание курьера и пробуем создать курьера только с логином и паролем"
+    )
+    def test_create_courier_partial_data(self, create_courier):
```
Нужно исправить: тут в тесте три комбинации параметров, а значит три теста должно быть

CRASH3000 replied:
✅ Переписал все тесты, где использовал параметризацию. Теперь на каждый параметр создал отдельный тест. Но опять же буду очень признателен за пояснение почему именно так нужно делать?

Так-так буквально на прошлом спринте нас просили писать тесты именно так, чтобы не разводить дублирущий код.

AlexKitIt replied:
Потому что ты перебираешь в тесте параметры и на каждый вызываешь assert. Если завалится на первом ты даже не узнаешь что там с остальными полями, тк тест дальше не пойдет. Про "убрать дублирование кода" это все так, и эти три теста стоит описать в одном параметризованном тестовом методе.

CRASH3000 committed:
API тесты для учебного сервиса Яндекс Самокат v3.0

Отработал все замечания ревьюера по API тестам.

Что нового?
- Создал файл в папке data для хранения тестовых данных.

Внес правки в код.
- Переписал логику, теперь при запуске каждого теста, создается новый курьер.
- Удалил assert во всех фикстурах.
- Добавил логику в фикстурах, где после завершения теста, созданный тестовый профиль курьера удаляется.
- Переписал все тесты, где использовал параметризацию. Теперь на каждый параметр создал отдельный тест.

- Удалил @pytest.mark.usefixtures из тестов
- Удалил предусловии из тестов
- Удалил дублирующие тесты и дублирующие проверки
- Удалил @allure.step из тестов

AlexKitIt commented on API-tests/test_01_create_courier_api.py:
```diff
@@ -0,0 +1,104 @@
+    @allure.description(
+        "Проверка, что курьера можно создать и ответом приходит код 201 и 'ok': True"
+    )
+    def test_create_courier(self, delete_courier):
```
Ты точно понимаешь как работает delete_courier? Фикстура возвращает список, в который по ее логике должен быть добавлен courier_id. И где этот параметр добавляется в этот список в тесте?

CRASH3000 replied:
✅  В тесте создания курьера я использую новую генерацию данных courier_data = generate_random_courier_data(), так как не могу напрямую использовать фикстуру create_courier. Если бы я использовал её, курьер уже был бы создан, и тест на создание провалился бы. Логика фикстуры была создана для других тестов, где для каждого создаётся новый курьер, а после завершения теста он удаляется. Я заметил свою ошибку и добавил в тест создания курьера логику добавления ID в список для удаления через delete_courier, так как до этого в список ничего не отправлялось, и курьер не удалялся.

AlexKitIt commented on API-tests/test_02_courier_login_api.py:
```diff
@@ -0,0 +1,68 @@
+        assert "id" in response.json(), "Ответ не содержит id"
+
+        # Сохранение ID курьера
+        self.__class__.courier_id = response.json()["id"]
```
Это зачем?

CRASH3000 replied:
✅ Спасибо, не увидел. Забыл удалить одну из своих старых задумок.

AlexKitIt commented on API-tests/conftest.py:
```diff
@@ -0,0 +1,50 @@
+    return courier_data
+
+@pytest.fixture(scope="function")
+def append_courier_id_for_delete(delete_courier, create_courier):
```
Эта фикстура нигде не используется

CRASH3000 replied:
✅  Я разделил логику на две части, чтобы следовать принципам “Чистого кода” и сделать код более модульным и гибким.

Логика заключалась в следующем:

- Фикстура create_courier отвечает исключительно за создание курьера и проверку успешности этого процесса. Это позволяет использовать её в разных контекстах, не привязываясь к логике удаления.

- Фикстура append_courier_id_for_delete добавляет ID курьера в список для последующего удаления. Этот подход позволяет отделить логику очистки данных (удаления) от создания, что повышает гибкость и снижает связанность кода.

Тем не менее, я понимаю вашу точку зрения, и чтобы упростить структуру, объединим логику создания и удаления в одну фикстуру, как вы предложили. Это облегчит понимание кода и сократит количество фикстур.

AlexKitIt commented on API-tests/conftest.py:
```diff
@@ -0,0 +1,50 @@
+
+
+@pytest.fixture(scope="function")
+def create_courier():
```
Вот здесь поставь вызов delete_courier а в теле фикстуры добавь получение courier_id и добавление его в список delete_courier. Тогда удаляться курьер у тебя будет безусловно даже если тест завалится, тк финализаторы фикстур отрабатывают в любом случае

CRASH3000 replied:
Спасибо, за свежий взгляд, действительно немного запутался во время реализации удаления курьера после тестов. 

✅  Правки внесены.

CRASH3000 committed:
API тесты для учебного сервиса Яндекс Самокат v4.0

Отработал все замечания ревьюера по API тестам.

Внес правки в код.
- Все однотипные тесты с разными параметрами переписал с использованием @pytest.mark.parametrize;
- Для тестов с параметризацией добавил динамичный Аллюр тайтл;
- Поправил тесты, где нужно создавать курьера прямо в тесте, реализовав механизм добавления ID в список для дальнейшего удаления курьера после теста (Пример теста: test_create_courier);
- Внес правки в фикстуру create_courier поставив вызов delete_courier;
- Перенес логику фикстуры append_courier_id_for_delete в create_courier;

- Удалил неиспользуемую логику в коде.

AlexKitIt commented on API-tests/test_01_create_courier_api.py:
```diff
@@ -0,0 +1,123 @@
+        assert response.status_code == 201
+        assert response.json() == expected_responses["courier"]["create_success"]
+
+        courier_id = response.json().get("id")
+        if courier_id:
+            delete_courier.append(courier_id)
```
Вот это всегда ставь до assert, так надежнее, если допустим изменится json ответа а курьер создастся - не удалится ничего

AlexKitIt commented on API-tests/test_01_create_courier_api.py:
```diff
@@ -0,0 +1,123 @@
+        assert response.json() == expected_responses["courier"]["create_success"]
+
+        courier_id = response.json().get("id")
+        if courier_id:
```
Про условия я уже писал, их не должно быть в тестах вообще. Всю логику мы описываем точно зная как это должно и будет работать
